lib/net.c

```diff
@@ -1555,29 +1583,72 @@ start_network_stage2(void *arg)
 	return ret;
 }
 
+static int
+drop_privileges(void)
+{
+	int ret;
+	char user[128];
+
+	ret = getlogin_r(user, sizeof(user));
+	if (ret != 0) {
+		G_LOG(ERR, "%s: failed to get the current logged in user - %s!\n",
+			__func__, strerror(errno));
+		return ret;
+	}
+
+	ret = initgroups(user, config.pw_gid);
+	if (ret != 0) {
+		G_LOG(ERR, "%s: failed to call initgrous(%s, %u) - %s!\n",
+			__func__, user, config.pw_gid, strerror(errno));
+		return ret;
+	}
+
+	/* Drop privileges. */
+	if (setgid(config.pw_gid) != 0 || setuid(config.pw_uid) != 0) {
+		G_LOG(ERR, "%s: failed to drop privileges for user with user id %u and group id %u: %s!\n",
+			__func__, config.pw_uid,
+			config.pw_gid, strerror(errno));
+		return -1;
+	}
+
+	return (seteuid(0) == -1 && setegid(0) == -1) ? 0 : -1;
+}
+
 int
-finalize_stage2(__attribute__((unused)) void *arg)
+finalize_stage2(void *arg)
 {
+	int ret;
+
 	if (ipv4_acl_enabled(&config.front)) {
-		int ret = build_ipv4_acls(&config.front);
+		ret = build_ipv4_acls(&config.front);
 		if (ret < 0)
 			return ret;
 	}
 	if (ipv4_acl_enabled(&config.back)) {
-		int ret = build_ipv4_acls(&config.back);
+		ret = build_ipv4_acls(&config.back);
 		if (ret < 0)
 			return ret;
 	}
 	if (ipv6_acl_enabled(&config.front)) {
-		int ret = build_ipv6_acls(&config.front);
+		ret = build_ipv6_acls(&config.front);
 		if (ret < 0)
 			return ret;
 	}
 	if (ipv6_acl_enabled(&config.back)) {
-		int ret = build_ipv6_acls(&config.back);
+		ret = build_ipv6_acls(&config.back);
 		if (ret < 0)
 			return ret;
 	}
+	if (config.pw_uid != 0) {
+		ret = fchown((intptr_t)arg, config.pw_uid, config.pw_gid);
```

Comment:
Let's avoid repeating the cast adding the variable: `int log_fd = (intptr_t)arg;`